src/haz3lcore/statics/TermBase.re

```diff
@@ -132,116 +137,124 @@ and Exp: {
     | OutsideAp;
 
   [@deriving (show({with_path: false}), sexp, yojson)]
-  type term =
+  type term('id) =
     | Invalid(string)
     | EmptyHole
     | MultiHole(list(Any.t))
-    | DynamicErrorHole(t, InvalidOperationError.t)
-    | FailedCast(t, Typ.t, Typ.t)
+    | DynamicErrorHole(t('id), InvalidOperationError.t)
+    | FailedCast(t('id), Typ.t, Typ.t)
     | Deferral(deferral_position)
     | Undefined
     | Bool(bool)
     | Int(int)
     | Float(float)
     | String(string)
-    | ListLit(list(t))
+    | ListLit(list(t('id)))
     | Constructor(string, Typ.t) // Typ.t field is only meaningful in dynamic expressions
     | Fun(
         Pat.t,
-        t,
+        t('id),
         [@show.opaque] option(ClosureEnvironment.t),
         option(Var.t),
       )
-    | TypFun(TPat.t, t, option(Var.t))
-    | Tuple(list(t))
+    | TypFun(TPat.t, t('id), option(Var.t))
+    | Tuple(list(t('id)))
     | Var(Var.t)
-    | Let(Pat.t, t, t)
-    | FixF(Pat.t, t, option(ClosureEnvironment.t))
-    | TyAlias(TPat.t, Typ.t, t)
-    | Ap(Operators.ap_direction, t, t)
-    | TypAp(t, Typ.t)
-    | DeferredAp(t, list(t))
-    | If(t, t, t)
-    | Seq(t, t)
-    | Test(t)
-    | Filter(StepperFilterKind.t, t)
-    | Closure([@show.opaque] ClosureEnvironment.t, t)
-    | Parens(t) // (
-    | Cons(t, t)
-    | ListConcat(t, t)
-    | UnOp(Operators.op_un, t)
-    | BinOp(Operators.op_bin, t, t)
+    | Let(Pat.t, t('id), t('id))
+    | FixF(Pat.t, t('id), option(ClosureEnvironment.t))
+    | TyAlias(TPat.t, Typ.t, t('id))
+    | Ap(Operators.ap_direction, t('id), t('id))
+    | TypAp(t('id), Typ.t)
+    | DeferredAp(t('id), list(t('id)))
+    | If(t('id), t('id), t('id))
+    | Seq(t('id), t('id))
+    | Test(t('id))
+    | Filter(StepperFilterKind.t, t('id))
+    | Closure([@show.opaque] ClosureEnvironment.t, t('id))
+    | Parens(t('id)) // (
+    | Cons(t('id), t('id))
+    | ListConcat(t('id), t('id))
+    | UnOp(Operators.op_un, t('id))
+    | BinOp(Operators.op_bin, t('id), t('id))
     | BuiltinFun(string)
-    | Match(t, list((Pat.t, t)))
+    | Match(t('id), list((Pat.t, t('id))))
     /* INVARIANT: in dynamic expressions, casts must be between
        two consistent types. Both types should be normalized in
        dynamics for the cast calculus to work right. */
-    | Cast(t, Typ.t, Typ.t) // first Typ.t field is only meaningful in dynamic expressions
-  and t = IdTagged.t(term);
+    | Cast(t('id), Typ.t, Typ.t) // first Typ.t field is only meaningful in dynamic expressions
+  and t('id) = IdTagged.t(term('id), 'id);
```

Comment:
This doesn't feel like the right encoding but I'd like to talk with someone else about it before I fall down any more rabbit holes